{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Jupyter Notebook UI to analyze baseline data from tap-habituation experiments!\n",
    "\n",
    "### Beginner Essentials:\n",
    "1. Shift-Enter to run each cell. After you run, you should see an output \"done step #\". If not, an error has occured\n",
    "2. When inputting your own code/revising the code, make sure you close all your quotation marks '' and brackets (), [], {}.\n",
    "3. Don't leave any commas (,) hanging! (make sure an object always follows a comma. If there is nothing after a comma, remove the comma!\n",
    "4. Learning to code? Each line of code is annotated to help you understand how this code works!\n",
    "\n",
    "**Run all cells/steps sequentially, even the ones that do not need input**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step-by-Step Analysis of the Jupyter Notebook\n",
    "\n",
    "| Step | Purpose | Key Actions |\n",
    "|------|---------|-------------|\n",
    "| **1. Import Packages** | Load required Python libraries for data analysis | Imports `pandas`, `numpy`, `matplotlib`, etc. | \n",
    "| **2. Pick Filepath** | Select the folder containing experimental data files (.dat or .trv) | Input required: Uses `FileChooser` widget to select directory | \n",
    "| **3. User-Defined Variables** | Set experiment parameters | Defines: `bin`  | \n",
    "| **4. Construct Filelist** | Find all files in selected folder | Sets working directory and scans `folder_path` using; Displays no. of `.trv` files found in the folder |\n",
    "| **5. Process Data Function** | Define functions to load, clean, and analyze raw data | - `ProcessData()`: Loads files, calculates metrics (reversal probability, speed) |\n",
    "| **6.1 Process Data** | Apply processing to all strains| - Checks `filelist` for unique strain names (e.g., \"N2\") <br>- Runs `ProcessData()` for each strain | \n",
    "| **7. Grouping & Naming** | Combine data from all strains | - Concatenates DataFrames<br>- Assigns dataset names (e.g., \"N2\") | \n",
    "| **Output CSV** | Save processed data | Exports `Baseline_data` to CSV |\n",
    "\n",
    "### Key Notes:\n",
    "- User Input Required: Steps 2 (file selection), 3 (parameters), 6.1 (strain verification)\n",
    "- Output: Final CSV contains all analyzed tap response data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Importing Packages Required (No input required, just run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done step 1\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd #<- package used to import and organize data\n",
    "import numpy as np #<- package used to import and organize data\n",
    "import seaborn as sns #<- package used to plot graphs\n",
    "from matplotlib import pyplot as plt #<- package used to plot graphs\n",
    "import os #<- package used to work with system filepaths\n",
    "from ipywidgets import widgets #<- widget tool to generate button\n",
    "from IPython.display import display #<- displays button\n",
    "from ipyfilechooser import FileChooser\n",
    "# from tkinter import Tk, filedialog #<- Tkinter is a GUI package\n",
    "from tqdm.notebook import tqdm\n",
    "# import dask.dataframe as dd\n",
    "# import pingouin as pg\n",
    "pd.set_option('display.max_columns', 50)\n",
    "print(\"done step 1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Pick filepath (just run and click button from output)\n",
    "\n",
    "Run the following cell and click the button 'Select Folder' to pick a filepath.\n",
    "\n",
    "**Important: Later on, this script uses the total file path for each file to import and group data. That means if your folder has whatever your strain is named, the script will not work.**\n",
    "\n",
    "(ex. if your folder has \"N2\" in it this script sees all files inside this folder as having the \"N2\" search key)\n",
    "\n",
    "**An easy fix is to just rename your folder to something else (make your strains lower-case, or just have the date)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1b59bc3f56c24b23b6e51a53c0847445",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FileChooser(path='/Users', filename='', title='', show_hidden=False, select_desc='Select', change_desc='Change…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "starting_directory = '/Users'\n",
    "chooser = FileChooser(starting_directory)\n",
    "display(chooser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022\n"
     ]
    }
   ],
   "source": [
    "print(chooser.selected_path)\n",
    "folder_path=chooser.selected_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e80363ff200a490084d9aa22e43fbe56",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Select(description='Screen:', options=('PD_Screen', 'ASD_Screen', 'G-Proteins_Screen', 'Glia_Genes_Screen', 'N…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "screens = ['PD_Screen', 'ASD_Screen', 'G-Proteins_Screen', 'Glia_Genes_Screen', 'Neuron_Genes_Screen']\n",
    "\n",
    "screen_chooser = widgets.Select(options=screens, value=screens[0], description='Screen:')\n",
    "display(screen_chooser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "Screen=screen_chooser.value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. User Defined Variables (Add input here)\n",
    "\n",
    "Here, we add some constants to help you blaze through this code.\n",
    "\n",
    "3.1: Setting time bins\n",
    "\n",
    "\n",
    "3.2: Setting view range for your graph\n",
    "- Top, bottom = y axis view range\n",
    "- left, right = x axis view range\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.000e+00 1.000e+00 2.000e+00 ... 1.198e+03 1.199e+03 1.200e+03]\n",
      "done step 3\n"
     ]
    }
   ],
   "source": [
    "# Setting 1s Bins\n",
    "bins = np.linspace(0,1200,1201) # np.linspace(start, end, steps in between)\n",
    "print(bins)\n",
    "\n",
    "\n",
    "print(\"done step 3\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Construct filelist from folder path (No input required, just run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of .dat files to process: 13\n",
      "done step 4\n"
     ]
    }
   ],
   "source": [
    "os.chdir(folder_path) # setting your working directory so that your images will be saved here\n",
    "\n",
    "filelist = list() # empty list\n",
    "for root, dirs, files in os.walk(folder_path): # this for loop goes through your folder \n",
    "    for name in files:\n",
    "        if name.endswith('.dat'): # and takes out all files with a .dat (file that contains your data)\n",
    "            if \"_\" in name.split(\".\")[-2]:\n",
    "                filepath = os.path.join(root, name) # Notes down the file path of each data file\n",
    "                filelist.append(filepath) # saves it into the list\n",
    "\n",
    "if not filelist:\n",
    "    raise FileNotFoundError(\"No .dat files found in the selected folder!\")\n",
    "else:\n",
    "    print(f\"Number of .dat files to process: {len(filelist)}\")\n",
    "    # print(f\"Example of first and last file saved: {filelist[0]}, {filelist[-1]}\") \n",
    "\n",
    "print('done step 4')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Process Data Function (No input required, just run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done step 5\n"
     ]
    }
   ],
   "source": [
    "def ProcessData(strain, experiment_counter): \n",
    "    \"\"\"\n",
    "    Filters and processes .dat files matching the given strain.\n",
    "\n",
    "    Parameters: \n",
    "        strain (str): keyword to match in the files\n",
    "\n",
    "    Returns:\n",
    "        dict: N (Plate number) and Dataframe with required columns \n",
    "              (\"time\", \"dura\", \"dist\", \"prob\", \"speed\", \"plate\", \"Date\",\n",
    "              \"Plate_id\", \"Screen\")\n",
    "\n",
    "    \"\"\"\n",
    "    strain_filelist = [x for x in filelist if strain in x] # Goes through the list and filters for keyword\n",
    "    Strain_N = len(strain_filelist) # Finds the number of plates per strain\n",
    "    if Strain_N == 0:\n",
    "        raise AssertionError ('{} is not a good identifier'.format(strain))\n",
    "    else:\n",
    "        pass\n",
    "        print(f'Strain {strain}')\n",
    "        print(f'Number of plates: {Strain_N}') \n",
    "        \n",
    "        # visiting files in this strain\n",
    "        strain_filelist = [file for file in filelist if strain in file]\n",
    "        df_list=[]\n",
    "        for i, file in enumerate(strain_filelist):\n",
    "            if file.split('/')[-1].startswith('._'):\n",
    "                pass\n",
    "            else:\n",
    "                try:\n",
    "                    print(f\"File: {file}\")\n",
    "                    df= pd.read_csv(file, sep=' ', header = None, encoding_errors='ignore')\n",
    "                    df['Plate_id'] = file.split('/')[-1].split('_')[-1].split('.')[0]\n",
    "                    df['Date'] = file.split('/')[-2].split('_')[0]\n",
    "                    df['Screen'] = file.split('/')[-4]\n",
    "                    df['Experiment'] = experiment_counter\n",
    "                    experiment_counter = 1+experiment_counter\n",
    "                    df_list.append(df)\n",
    "                except:\n",
    "                    print(f\"error in file {file}\")\n",
    "                    pass\n",
    "        DF_Total = pd.concat(df_list, ignore_index = True)\n",
    "        DF_Total = DF_Total.rename( \n",
    "                    {0:'Time',\n",
    "                    1:'n',\n",
    "                    2:'Number',\n",
    "                    3:'Instantaneous Speed',\n",
    "                    4:'Interval Speed',\n",
    "                    5:'Bias',\n",
    "                    6:'Tap',\n",
    "                    7:'Puff',\n",
    "                    8:'x',\n",
    "                    9:'y',\n",
    "                    10:'Morphwidth',\n",
    "                    11:'Midline',\n",
    "                    12:'Area',\n",
    "                    13:'Angular Speed',\n",
    "                    14:'Aspect Ratio',\n",
    "                    15:'Kink',\n",
    "                    16:'Curve',\n",
    "                    17:'Crab',\n",
    "                    18:'Pathlength'}, axis=1)\n",
    "        \n",
    "        # check function here for NaN Columns\n",
    "        DF_Total['plate'] = 0\n",
    "\n",
    "        print(\"---------------------------------------------------------------------------------------------------------------------------------------------------------------------------\")\n",
    "\n",
    "    return{\n",
    "            'N': Strain_N,\n",
    "            'Confirm':DF_Total,\n",
    "            'experiment_counter': experiment_counter\n",
    "            # 'Final': DF_Final\n",
    "    }\n",
    "\n",
    "\n",
    "print('done step 5')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6.1 Process Data\n",
    "\n",
    "Create a dictionary `StrainNames` that contains all the genotype/strain names from each file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genotypes/strains in the experiment: 3\n",
      "1: N2\n",
      "2: hipr-1_ok1081\n",
      "3: hipr-1_tm10120\n",
      "done step 6.1\n"
     ]
    }
   ],
   "source": [
    "genotype=[]\n",
    "for f in filelist:\n",
    "    genotype.append(f.split('/')[-3])\n",
    "\n",
    "genotypes=np.unique(genotype).tolist()\n",
    "\n",
    "if Screen ==\"Neuron_Genes_Screen\":\n",
    "    genotypes.insert(0, genotypes.pop(genotypes.index(\"N2_XJ1\")))\n",
    "    genotypes.insert(0, genotypes.pop(genotypes.index(\"N2_N2\")))\n",
    "else:\n",
    "    genotypes.insert(0, genotypes.pop(genotypes.index(\"N2\")))\n",
    "\n",
    "nstrains = list(range(1, len(genotypes) + 1))\n",
    "StrainNames = {nstrains[i]: genotypes[i] for i in range(len(nstrains))}\n",
    "\n",
    "print(f\"Number of genotypes/strains in the experiment: {len(genotypes)}\")\n",
    "\n",
    "# Display the first 5 Strain names in the experiment\n",
    "for k in list(StrainNames)[:5]:\n",
    "    print(f\"{k}: {StrainNames[k]}\")\n",
    "\n",
    "\n",
    "print(\"done step 6.1\")\n",
    "\n",
    "# <---------------- Test element to use for dictionary buidling -------------------\n",
    "# s = '/Users/Joseph/Desktop/OnFoodOffFoodTest/N2_OnFood/20220401_163048/N2_10x1_n96h20C_360sA0401_ka.00065.dat'\n",
    "# slist=s.split('/')[5]\n",
    "# print(slist)\n",
    "# print(list(range(1,5+1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6.2 Process Data (just run this cell)\n",
    "\n",
    "Pass each strain through `ProcessData()` function "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bad401c676f54549bf57463498e974eb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strain N2\n",
      "Number of plates: 5\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/N2/20220815_101538/N2_10x2_f72h20C_600s31x10s10s_B0811ab.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/N2/20220815_102652/N2_10x2_f96h20C_600s31x10s10s_A0811aa.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/N2/20220815_122801/N2_10x2_f96h20C_600s31x10s10s_A0811ad.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/N2/20220815_121502/N2_10x2_f72h20C_600s31x10s10s_B0811ae.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/N2/20220815_103433/N2_10x2_f72h20C_600s31x10s10s_C0811ac.dat\n",
      "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "Strain hipr-1_ok1081\n",
      "Number of plates: 4\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_ok1081/20220815_114829_removed/RB1102_10x2_f96h20C_600s31x10s10s_A0811bc.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_ok1081/20220815_100526/RB1102_10x2_f72h20C_600s31x10s10s_C0811bb.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_ok1081/20220815_114931/RB1102_10x2_f72h20C_600s31x10s10s_B0811bd.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_ok1081/20220815_120440/RB1102_10x2_f72h20C_600s31x10s10s_C0811be.dat\n",
      "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "Strain hipr-1_tm10120\n",
      "Number of plates: 4\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_tm10120/20220815_111005/FX31344_10x2_f72h20C_600s31x10s10s_B0811cc.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_tm10120/20220815_113906_removed/FX31344_10x2_f72h20C_600s31x10s10s_C0811cd.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_tm10120/20220815_093805/FX31344_10x2_f72h20C_600s31x10s10s_C0811ca.dat\n",
      "File: /Users/gurmehak/Documents/RankinLab/Test_Datasets/PDScreen_TapHab_August15_2022/hipr-1_tm10120/20220815_112320/FX31344_10x2_f96h20C_600s31x10s10s_A0811cb.dat\n",
      "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "done step 6.2\n"
     ]
    }
   ],
   "source": [
    "DataLists = [0] # generates empty list at index 0 because we want indexing to start at 1 \n",
    "                # when I say #1, I want the first point, not the second point\n",
    "\n",
    "experiment_counter = 1\n",
    "\n",
    "# the loop below goes through the dictionary in step 6.1 and processes data\n",
    "# and appends all data into a list of dataframes\n",
    "for s in tqdm(StrainNames.values()): \n",
    "    if not s == '':\n",
    "        result = ProcessData(s, experiment_counter)\n",
    "        DataLists.append(result['Confirm'])\n",
    "        experiment_counter = result['experiment_counter'] \n",
    "\n",
    "print('done step 6.2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert float64 data to float32 to reduce memory load (can also convert to 16 if needed)\n",
    "\n",
    "For plain english:\n",
    "\n",
    "float16 = 4 decimal points\n",
    "\n",
    "float32 = 8 decimal points\n",
    "\n",
    "float64 = 16 decimal points\n",
    "\n",
    "more decimal points = more data/memory that computer has to keep track of"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# commented out this section in case memory load needs to be reduced\n",
    "\n",
    "# for n in tqdm(DataLists[1:]:\n",
    "#     print(n)\n",
    "#     TestData=n\n",
    "#     TestData[TestData.select_dtypes(np.float64).columns]=TestData.select_dtypes(np.float64).astype(np.float16)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Grouping Data and Naming\n",
    "\n",
    "This step takes all the individual strain data (processed in Step 6) and combines them into single dataframe, filters for time window 490s - 590s, drops unwanted columns. \n",
    "\n",
    "The final processed data `Baseline_data` is ready for analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Time</th>\n",
       "      <th>n</th>\n",
       "      <th>Number</th>\n",
       "      <th>Instantaneous Speed</th>\n",
       "      <th>Interval Speed</th>\n",
       "      <th>Bias</th>\n",
       "      <th>Morphwidth</th>\n",
       "      <th>Midline</th>\n",
       "      <th>Area</th>\n",
       "      <th>Angular Speed</th>\n",
       "      <th>Aspect Ratio</th>\n",
       "      <th>Kink</th>\n",
       "      <th>Curve</th>\n",
       "      <th>Crab</th>\n",
       "      <th>Pathlength</th>\n",
       "      <th>Plate_id</th>\n",
       "      <th>Date</th>\n",
       "      <th>Screen</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9775</th>\n",
       "      <td>490.016</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0823</td>\n",
       "      <td>0.1195</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.1078</td>\n",
       "      <td>1.0908</td>\n",
       "      <td>0.142641</td>\n",
       "      <td>6.1</td>\n",
       "      <td>0.225</td>\n",
       "      <td>31.1</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0066</td>\n",
       "      <td>11.893</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9776</th>\n",
       "      <td>490.056</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>0.1024</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.1059</td>\n",
       "      <td>1.0880</td>\n",
       "      <td>0.140150</td>\n",
       "      <td>5.3</td>\n",
       "      <td>0.219</td>\n",
       "      <td>30.9</td>\n",
       "      <td>24.7</td>\n",
       "      <td>0.0064</td>\n",
       "      <td>11.896</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9777</th>\n",
       "      <td>490.103</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0784</td>\n",
       "      <td>0.1024</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.1050</td>\n",
       "      <td>1.0914</td>\n",
       "      <td>0.138935</td>\n",
       "      <td>5.2</td>\n",
       "      <td>0.219</td>\n",
       "      <td>31.0</td>\n",
       "      <td>24.6</td>\n",
       "      <td>0.0057</td>\n",
       "      <td>11.898</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9778</th>\n",
       "      <td>490.144</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0970</td>\n",
       "      <td>0.1118</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.1054</td>\n",
       "      <td>1.0935</td>\n",
       "      <td>0.140575</td>\n",
       "      <td>5.8</td>\n",
       "      <td>0.218</td>\n",
       "      <td>30.6</td>\n",
       "      <td>24.5</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>11.901</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9779</th>\n",
       "      <td>490.186</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0994</td>\n",
       "      <td>0.1197</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.1110</td>\n",
       "      <td>1.1026</td>\n",
       "      <td>0.146894</td>\n",
       "      <td>5.6</td>\n",
       "      <td>0.227</td>\n",
       "      <td>31.4</td>\n",
       "      <td>24.4</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>11.904</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Time   n  Number  Instantaneous Speed  Interval Speed  Bias  \\\n",
       "9775  490.016  14      12               0.0823          0.1195  0.25   \n",
       "9776  490.056  14      12               0.0736          0.1024  0.25   \n",
       "9777  490.103  14      12               0.0784          0.1024  0.25   \n",
       "9778  490.144  14      12               0.0970          0.1118  0.25   \n",
       "9779  490.186  14      12               0.0994          0.1197  0.25   \n",
       "\n",
       "      Morphwidth  Midline      Area  Angular Speed  Aspect Ratio  Kink  Curve  \\\n",
       "9775      0.1078   1.0908  0.142641            6.1         0.225  31.1   25.0   \n",
       "9776      0.1059   1.0880  0.140150            5.3         0.219  30.9   24.7   \n",
       "9777      0.1050   1.0914  0.138935            5.2         0.219  31.0   24.6   \n",
       "9778      0.1054   1.0935  0.140575            5.8         0.218  30.6   24.5   \n",
       "9779      0.1110   1.1026  0.146894            5.6         0.227  31.4   24.4   \n",
       "\n",
       "        Crab  Pathlength Plate_id      Date     Screen dataset Gene Allele  \n",
       "9775  0.0066      11.893  B0811ab  20220815  PD_Screen      N2   N2     N2  \n",
       "9776  0.0064      11.896  B0811ab  20220815  PD_Screen      N2   N2     N2  \n",
       "9777  0.0057      11.898  B0811ab  20220815  PD_Screen      N2   N2     N2  \n",
       "9778  0.0103      11.901  B0811ab  20220815  PD_Screen      N2   N2     N2  \n",
       "9779  0.0097      11.904  B0811ab  20220815  PD_Screen      N2   N2     N2  "
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Baseline_data=pd.concat(df.assign(dataset=StrainNames.get(i+1)) for i, df in enumerate(DataLists[1:]))\n",
    "\n",
    "Baseline_data[['Gene', 'Allele']] = Baseline_data['dataset'].str.split('_', n=1, expand=True)\n",
    "\n",
    "Baseline_data['Allele'] = Baseline_data['Allele'].fillna('N2')\n",
    "\n",
    "Baseline_data = Baseline_data.drop(columns=[\"plate\", \"Tap\", \"Puff\", \"x\",\"y\", \"Experiment\"]).dropna().reset_index(drop=True)\n",
    "\n",
    "Baseline_data['Screen']=Screen\n",
    "\n",
    "Baseline_data = Baseline_data[((Baseline_data.Time<=590)&(Baseline_data.Time >=490))] \n",
    "\n",
    "Baseline_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30487, 21)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Baseline_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating Post Stimulus Data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "# similar filters as baseline data\n",
    "\n",
    "Post_stimulus_data_pre=pd.concat(df.assign(dataset=StrainNames.get(i+1)) for i,df in enumerate(DataLists[1:]))\n",
    "\n",
    "Post_stimulus_data_pre[['Gene', 'Allele']] = Post_stimulus_data_pre['dataset'].str.split('_', n=1, expand=True)\n",
    "\n",
    "Post_stimulus_data_pre['Allele'] = Post_stimulus_data_pre['Allele'].fillna('N2')\n",
    "\n",
    "Post_stimulus_data_pre = Post_stimulus_data_pre.drop(columns=[\"plate\", \"Puff\", \"x\",\"y\"]).dropna().reset_index(drop=True)\n",
    "\n",
    "Post_stimulus_data_pre['Screen']=Screen\n",
    "\n",
    "Post_stimulus_data_pre = Post_stimulus_data_pre[((Post_stimulus_data_pre.Time>599))]\n",
    "\n",
    "Post_stimulus_data_pre['Time'] = round(Post_stimulus_data_pre['Time']).astype('int')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Time</th>\n",
       "      <th>n</th>\n",
       "      <th>Number</th>\n",
       "      <th>Instantaneous Speed</th>\n",
       "      <th>Interval Speed</th>\n",
       "      <th>Bias</th>\n",
       "      <th>Tap</th>\n",
       "      <th>Morphwidth</th>\n",
       "      <th>Midline</th>\n",
       "      <th>Area</th>\n",
       "      <th>Angular Speed</th>\n",
       "      <th>Aspect Ratio</th>\n",
       "      <th>Kink</th>\n",
       "      <th>Curve</th>\n",
       "      <th>Crab</th>\n",
       "      <th>Pathlength</th>\n",
       "      <th>Plate_id</th>\n",
       "      <th>Date</th>\n",
       "      <th>Screen</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Tap_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12094</th>\n",
       "      <td>599</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>0.0692</td>\n",
       "      <td>0.1084</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>1.0889</td>\n",
       "      <td>0.139812</td>\n",
       "      <td>3.7</td>\n",
       "      <td>0.249</td>\n",
       "      <td>38.8</td>\n",
       "      <td>27.6</td>\n",
       "      <td>0.0067</td>\n",
       "      <td>6.614</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12095</th>\n",
       "      <td>599</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>0.0885</td>\n",
       "      <td>0.1161</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1046</td>\n",
       "      <td>1.0857</td>\n",
       "      <td>0.139187</td>\n",
       "      <td>4.8</td>\n",
       "      <td>0.247</td>\n",
       "      <td>37.1</td>\n",
       "      <td>27.4</td>\n",
       "      <td>0.0085</td>\n",
       "      <td>6.618</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12096</th>\n",
       "      <td>599</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>0.0967</td>\n",
       "      <td>0.1244</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1059</td>\n",
       "      <td>1.0914</td>\n",
       "      <td>0.141114</td>\n",
       "      <td>5.2</td>\n",
       "      <td>0.250</td>\n",
       "      <td>39.6</td>\n",
       "      <td>27.6</td>\n",
       "      <td>0.0092</td>\n",
       "      <td>6.622</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12097</th>\n",
       "      <td>599</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1135</td>\n",
       "      <td>0.1604</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1103</td>\n",
       "      <td>1.1059</td>\n",
       "      <td>0.146321</td>\n",
       "      <td>5.1</td>\n",
       "      <td>0.253</td>\n",
       "      <td>38.1</td>\n",
       "      <td>27.5</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>6.626</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12098</th>\n",
       "      <td>599</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1131</td>\n",
       "      <td>0.1630</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1059</td>\n",
       "      <td>1.0967</td>\n",
       "      <td>0.141478</td>\n",
       "      <td>4.9</td>\n",
       "      <td>0.248</td>\n",
       "      <td>37.9</td>\n",
       "      <td>27.4</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>6.632</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Time   n  Number  Instantaneous Speed  Interval Speed   Bias  Tap  \\\n",
       "12094   599  16      14               0.0692          0.1084  0.429    0   \n",
       "12095   599  16      14               0.0885          0.1161  0.429    0   \n",
       "12096   599  16      14               0.0967          0.1244  0.429    0   \n",
       "12097   599  16      14               0.1135          0.1604  0.571    0   \n",
       "12098   599  16      14               0.1131          0.1630  0.571    0   \n",
       "\n",
       "       Morphwidth  Midline      Area  Angular Speed  Aspect Ratio  Kink  \\\n",
       "12094      0.1045   1.0889  0.139812            3.7         0.249  38.8   \n",
       "12095      0.1046   1.0857  0.139187            4.8         0.247  37.1   \n",
       "12096      0.1059   1.0914  0.141114            5.2         0.250  39.6   \n",
       "12097      0.1103   1.1059  0.146321            5.1         0.253  38.1   \n",
       "12098      0.1059   1.0967  0.141478            4.9         0.248  37.9   \n",
       "\n",
       "       Curve    Crab  Pathlength Plate_id      Date     Screen  Experiment  \\\n",
       "12094   27.6  0.0067       6.614  B0811ab  20220815  PD_Screen           1   \n",
       "12095   27.4  0.0085       6.618  B0811ab  20220815  PD_Screen           1   \n",
       "12096   27.6  0.0092       6.622  B0811ab  20220815  PD_Screen           1   \n",
       "12097   27.5  0.0094       6.626  B0811ab  20220815  PD_Screen           1   \n",
       "12098   27.4  0.0111       6.632  B0811ab  20220815  PD_Screen           1   \n",
       "\n",
       "      dataset Gene Allele  Tap_num  \n",
       "12094      N2   N2     N2        0  \n",
       "12095      N2   N2     N2        0  \n",
       "12096      N2   N2     N2        0  \n",
       "12097      N2   N2     N2        0  \n",
       "12098      N2   N2     N2        0  "
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add continuous tap numbers from 1 to 31 for each experiment\n",
    "# E.g., Experiment 1 has taps 1-31, Experiment 2 has taps 1-31 and so on..\n",
    "\n",
    "Post_stimulus_data_pre['Tap_num'] = Post_stimulus_data_pre.groupby(['Experiment'])['Tap'].cumsum()\n",
    "\n",
    "Post_stimulus_data_pre.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Time</th>\n",
       "      <th>n</th>\n",
       "      <th>Number</th>\n",
       "      <th>Instantaneous Speed</th>\n",
       "      <th>Interval Speed</th>\n",
       "      <th>Bias</th>\n",
       "      <th>Tap</th>\n",
       "      <th>Morphwidth</th>\n",
       "      <th>Midline</th>\n",
       "      <th>Area</th>\n",
       "      <th>Angular Speed</th>\n",
       "      <th>Aspect Ratio</th>\n",
       "      <th>Kink</th>\n",
       "      <th>Curve</th>\n",
       "      <th>Crab</th>\n",
       "      <th>Pathlength</th>\n",
       "      <th>Plate_id</th>\n",
       "      <th>Date</th>\n",
       "      <th>Screen</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Tap_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12270</th>\n",
       "      <td>607</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1634</td>\n",
       "      <td>0.1742</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1153</td>\n",
       "      <td>1.0888</td>\n",
       "      <td>0.145435</td>\n",
       "      <td>7.5</td>\n",
       "      <td>0.380</td>\n",
       "      <td>56.0</td>\n",
       "      <td>39.2</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>6.378</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12271</th>\n",
       "      <td>607</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1317</td>\n",
       "      <td>0.1498</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1132</td>\n",
       "      <td>1.0722</td>\n",
       "      <td>0.142155</td>\n",
       "      <td>7.6</td>\n",
       "      <td>0.382</td>\n",
       "      <td>54.8</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>6.387</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12272</th>\n",
       "      <td>607</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1287</td>\n",
       "      <td>0.1541</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1182</td>\n",
       "      <td>1.0861</td>\n",
       "      <td>0.148143</td>\n",
       "      <td>6.3</td>\n",
       "      <td>0.380</td>\n",
       "      <td>57.8</td>\n",
       "      <td>38.8</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>6.397</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12273</th>\n",
       "      <td>607</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1197</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1147</td>\n",
       "      <td>1.0684</td>\n",
       "      <td>0.143301</td>\n",
       "      <td>5.6</td>\n",
       "      <td>0.370</td>\n",
       "      <td>56.5</td>\n",
       "      <td>37.9</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>6.406</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12274</th>\n",
       "      <td>607</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>0.1149</td>\n",
       "      <td>0.1522</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1112</td>\n",
       "      <td>1.0703</td>\n",
       "      <td>0.139916</td>\n",
       "      <td>6.9</td>\n",
       "      <td>0.375</td>\n",
       "      <td>54.5</td>\n",
       "      <td>37.7</td>\n",
       "      <td>0.0136</td>\n",
       "      <td>6.415</td>\n",
       "      <td>B0811ab</td>\n",
       "      <td>20220815</td>\n",
       "      <td>PD_Screen</td>\n",
       "      <td>1</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Time   n  Number  Instantaneous Speed  Interval Speed   Bias  Tap  \\\n",
       "12270   607  17      14               0.1634          0.1742  0.500    0   \n",
       "12271   607  17      14               0.1317          0.1498  0.500    0   \n",
       "12272   607  17      14               0.1287          0.1541  0.500    0   \n",
       "12273   607  17      14               0.1197          0.1546  0.571    0   \n",
       "12274   607  17      14               0.1149          0.1522  0.571    0   \n",
       "\n",
       "       Morphwidth  Midline      Area  Angular Speed  Aspect Ratio  Kink  \\\n",
       "12270      0.1153   1.0888  0.145435            7.5         0.380  56.0   \n",
       "12271      0.1132   1.0722  0.142155            7.6         0.382  54.8   \n",
       "12272      0.1182   1.0861  0.148143            6.3         0.380  57.8   \n",
       "12273      0.1147   1.0684  0.143301            5.6         0.370  56.5   \n",
       "12274      0.1112   1.0703  0.139916            6.9         0.375  54.5   \n",
       "\n",
       "       Curve    Crab  Pathlength Plate_id      Date     Screen  Experiment  \\\n",
       "12270   39.2  0.0156       6.378  B0811ab  20220815  PD_Screen           1   \n",
       "12271   39.0  0.0156       6.387  B0811ab  20220815  PD_Screen           1   \n",
       "12272   38.8  0.0154       6.397  B0811ab  20220815  PD_Screen           1   \n",
       "12273   37.9  0.0164       6.406  B0811ab  20220815  PD_Screen           1   \n",
       "12274   37.7  0.0136       6.415  B0811ab  20220815  PD_Screen           1   \n",
       "\n",
       "      dataset Gene Allele  Tap_num  \n",
       "12270      N2   N2     N2        1  \n",
       "12271      N2   N2     N2        1  \n",
       "12272      N2   N2     N2        1  \n",
       "12273      N2   N2     N2        1  \n",
       "12274      N2   N2     N2        1  "
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create windows from 7s to 9.5s post a tap (\"Tap\"=1) for each experiment\n",
    "# and concatenate all these wondows into a single dataframe\n",
    "\n",
    "Post_stimulus_data = []\n",
    "\n",
    "for exp in Post_stimulus_data_pre['Experiment'].unique(): # loop through each experiment separately \n",
    "    df = Post_stimulus_data_pre[Post_stimulus_data_pre['Experiment'] == exp]  \n",
    "    tap_times = df[df['Tap'] == 1]['Time']  # get times where tap occured\n",
    "\n",
    "    for t in tap_times: \n",
    "        window = df[(df['Time'] >= t + 7) & (df['Time'] <= t + 9.5)]\n",
    "        Post_stimulus_data.append(window)\n",
    "\n",
    "Post_stimulus_data = pd.concat(Post_stimulus_data)\n",
    "\n",
    "Post_stimulus_data.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Tap_num</th>\n",
       "      <th>Time</th>\n",
       "      <th>n</th>\n",
       "      <th>Number</th>\n",
       "      <th>Instantaneous Speed</th>\n",
       "      <th>Interval Speed</th>\n",
       "      <th>Bias</th>\n",
       "      <th>Tap</th>\n",
       "      <th>Morphwidth</th>\n",
       "      <th>Midline</th>\n",
       "      <th>Area</th>\n",
       "      <th>Angular Speed</th>\n",
       "      <th>Aspect Ratio</th>\n",
       "      <th>Kink</th>\n",
       "      <th>Curve</th>\n",
       "      <th>Crab</th>\n",
       "      <th>Pathlength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>607</td>\n",
       "      <td>16.925000</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>0.125645</td>\n",
       "      <td>0.135265</td>\n",
       "      <td>0.350025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.112155</td>\n",
       "      <td>1.072315</td>\n",
       "      <td>0.141810</td>\n",
       "      <td>7.340000</td>\n",
       "      <td>0.344050</td>\n",
       "      <td>48.107500</td>\n",
       "      <td>36.130000</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>6.499075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>617</td>\n",
       "      <td>16.972973</td>\n",
       "      <td>14.891892</td>\n",
       "      <td>0.214500</td>\n",
       "      <td>0.144012</td>\n",
       "      <td>0.527905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.111914</td>\n",
       "      <td>1.051191</td>\n",
       "      <td>0.137795</td>\n",
       "      <td>16.145946</td>\n",
       "      <td>0.315459</td>\n",
       "      <td>53.440541</td>\n",
       "      <td>32.040541</td>\n",
       "      <td>0.026011</td>\n",
       "      <td>6.607365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>627</td>\n",
       "      <td>16.629032</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>0.235965</td>\n",
       "      <td>0.107100</td>\n",
       "      <td>0.477129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.109982</td>\n",
       "      <td>1.079510</td>\n",
       "      <td>0.140502</td>\n",
       "      <td>12.866129</td>\n",
       "      <td>0.293274</td>\n",
       "      <td>51.256452</td>\n",
       "      <td>30.583871</td>\n",
       "      <td>0.025710</td>\n",
       "      <td>7.168435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>637</td>\n",
       "      <td>13.134328</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>0.235875</td>\n",
       "      <td>0.126124</td>\n",
       "      <td>0.965970</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.105619</td>\n",
       "      <td>1.042824</td>\n",
       "      <td>0.133841</td>\n",
       "      <td>20.716418</td>\n",
       "      <td>0.333090</td>\n",
       "      <td>58.326866</td>\n",
       "      <td>31.059701</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>8.156448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>647</td>\n",
       "      <td>16.108108</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>0.266666</td>\n",
       "      <td>0.124458</td>\n",
       "      <td>0.958500</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.103418</td>\n",
       "      <td>1.075350</td>\n",
       "      <td>0.137081</td>\n",
       "      <td>13.786486</td>\n",
       "      <td>0.281851</td>\n",
       "      <td>39.606757</td>\n",
       "      <td>29.005405</td>\n",
       "      <td>0.027261</td>\n",
       "      <td>9.523203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>13</td>\n",
       "      <td>27</td>\n",
       "      <td>867</td>\n",
       "      <td>24.480000</td>\n",
       "      <td>21.626667</td>\n",
       "      <td>0.150065</td>\n",
       "      <td>0.086068</td>\n",
       "      <td>0.895293</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.089453</td>\n",
       "      <td>0.974081</td>\n",
       "      <td>0.110356</td>\n",
       "      <td>6.172000</td>\n",
       "      <td>0.238000</td>\n",
       "      <td>40.257333</td>\n",
       "      <td>26.874667</td>\n",
       "      <td>0.011888</td>\n",
       "      <td>14.472000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>13</td>\n",
       "      <td>28</td>\n",
       "      <td>877</td>\n",
       "      <td>23.120000</td>\n",
       "      <td>21.040000</td>\n",
       "      <td>0.148219</td>\n",
       "      <td>0.085596</td>\n",
       "      <td>0.856187</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088217</td>\n",
       "      <td>0.962596</td>\n",
       "      <td>0.108167</td>\n",
       "      <td>4.645333</td>\n",
       "      <td>0.218173</td>\n",
       "      <td>35.217333</td>\n",
       "      <td>25.397333</td>\n",
       "      <td>0.009620</td>\n",
       "      <td>12.475093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>13</td>\n",
       "      <td>29</td>\n",
       "      <td>887</td>\n",
       "      <td>23.391892</td>\n",
       "      <td>22.391892</td>\n",
       "      <td>0.129711</td>\n",
       "      <td>0.077372</td>\n",
       "      <td>0.782405</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091149</td>\n",
       "      <td>0.934377</td>\n",
       "      <td>0.105980</td>\n",
       "      <td>5.264865</td>\n",
       "      <td>0.217757</td>\n",
       "      <td>35.762162</td>\n",
       "      <td>25.420270</td>\n",
       "      <td>0.009753</td>\n",
       "      <td>12.697027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>13</td>\n",
       "      <td>30</td>\n",
       "      <td>897</td>\n",
       "      <td>21.547945</td>\n",
       "      <td>20.547945</td>\n",
       "      <td>0.136551</td>\n",
       "      <td>0.079448</td>\n",
       "      <td>0.855685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.086688</td>\n",
       "      <td>0.952897</td>\n",
       "      <td>0.105771</td>\n",
       "      <td>4.691781</td>\n",
       "      <td>0.212603</td>\n",
       "      <td>37.989041</td>\n",
       "      <td>25.341096</td>\n",
       "      <td>0.009581</td>\n",
       "      <td>12.224658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>13</td>\n",
       "      <td>31</td>\n",
       "      <td>1197</td>\n",
       "      <td>23.256757</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>0.071120</td>\n",
       "      <td>0.066065</td>\n",
       "      <td>0.372135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.090659</td>\n",
       "      <td>0.956012</td>\n",
       "      <td>0.108283</td>\n",
       "      <td>6.883784</td>\n",
       "      <td>0.351730</td>\n",
       "      <td>58.454054</td>\n",
       "      <td>32.710811</td>\n",
       "      <td>0.009481</td>\n",
       "      <td>22.026554</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>402 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Experiment  Tap_num  Time          n     Number  Instantaneous Speed  \\\n",
       "0             1        1   607  16.925000  14.000000             0.125645   \n",
       "1             1        2   617  16.972973  14.891892             0.214500   \n",
       "2             1        3   627  16.629032  13.000000             0.235965   \n",
       "3             1        4   637  13.134328  12.000000             0.235875   \n",
       "4             1        5   647  16.108108  12.000000             0.266666   \n",
       "..          ...      ...   ...        ...        ...                  ...   \n",
       "397          13       27   867  24.480000  21.626667             0.150065   \n",
       "398          13       28   877  23.120000  21.040000             0.148219   \n",
       "399          13       29   887  23.391892  22.391892             0.129711   \n",
       "400          13       30   897  21.547945  20.547945             0.136551   \n",
       "401          13       31  1197  23.256757  17.000000             0.071120   \n",
       "\n",
       "     Interval Speed      Bias  Tap  Morphwidth   Midline      Area  \\\n",
       "0          0.135265  0.350025  0.0    0.112155  1.072315  0.141810   \n",
       "1          0.144012  0.527905  0.0    0.111914  1.051191  0.137795   \n",
       "2          0.107100  0.477129  0.0    0.109982  1.079510  0.140502   \n",
       "3          0.126124  0.965970  0.0    0.105619  1.042824  0.133841   \n",
       "4          0.124458  0.958500  0.0    0.103418  1.075350  0.137081   \n",
       "..              ...       ...  ...         ...       ...       ...   \n",
       "397        0.086068  0.895293  0.0    0.089453  0.974081  0.110356   \n",
       "398        0.085596  0.856187  0.0    0.088217  0.962596  0.108167   \n",
       "399        0.077372  0.782405  0.0    0.091149  0.934377  0.105980   \n",
       "400        0.079448  0.855685  0.0    0.086688  0.952897  0.105771   \n",
       "401        0.066065  0.372135  0.0    0.090659  0.956012  0.108283   \n",
       "\n",
       "     Angular Speed  Aspect Ratio       Kink      Curve      Crab  Pathlength  \n",
       "0         7.340000      0.344050  48.107500  36.130000  0.015625    6.499075  \n",
       "1        16.145946      0.315459  53.440541  32.040541  0.026011    6.607365  \n",
       "2        12.866129      0.293274  51.256452  30.583871  0.025710    7.168435  \n",
       "3        20.716418      0.333090  58.326866  31.059701  0.026849    8.156448  \n",
       "4        13.786486      0.281851  39.606757  29.005405  0.027261    9.523203  \n",
       "..             ...           ...        ...        ...       ...         ...  \n",
       "397       6.172000      0.238000  40.257333  26.874667  0.011888   14.472000  \n",
       "398       4.645333      0.218173  35.217333  25.397333  0.009620   12.475093  \n",
       "399       5.264865      0.217757  35.762162  25.420270  0.009753   12.697027  \n",
       "400       4.691781      0.212603  37.989041  25.341096  0.009581   12.224658  \n",
       "401       6.883784      0.351730  58.454054  32.710811  0.009481   22.026554  \n",
       "\n",
       "[402 rows x 18 columns]"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregate columns by \"Experiment\" + \"Tap_num\" by taking their means\n",
    "\n",
    "Post_stimulus_data = Post_stimulus_data.groupby(['Experiment', 'Tap_num']).agg({\n",
    "    'Time': 'min', # take minimum valu of time instead of mean\n",
    "    'n': 'mean',\n",
    "    'Number': 'mean',\n",
    "    'Instantaneous Speed': 'mean',\n",
    "    'Interval Speed' : 'mean',\n",
    "    'Bias': 'mean',\n",
    "    'Tap': 'mean',\n",
    "    'Morphwidth': 'mean',\n",
    "    'Midline': 'mean',\n",
    "    'Area': 'mean',\n",
    "    'Angular Speed': 'mean',\n",
    "    'Aspect Ratio': 'mean',\n",
    "    'Kink': 'mean',\n",
    "    'Curve': 'mean',\n",
    "    'Crab': 'mean',\n",
    "    'Pathlength': 'mean'\n",
    "})\n",
    "\n",
    "Post_stimulus_data = Post_stimulus_data.reset_index()\n",
    "\n",
    "Post_stimulus_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done step 7\n"
     ]
    }
   ],
   "source": [
    "print('done step 7')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save dataframe as `.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved as .csv!\n"
     ]
    }
   ],
   "source": [
    "Baseline_data.to_csv(f\"{Baseline_data.Screen[0]}_baseline_output.csv\")\n",
    "print('saved Baseline data as .csv!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Post_stimulus_data.to_csv(f\"{Post_stimulus_data.Screen[0]}_post_stimulus.csv\")\n",
    "print('saved Post stimulus data as .csv!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Done!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rankinlab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
